#    Description                  : stripPointerCastsAndOffsets should not process across AddrSpaceCast


diff -Naur --strip-trailing-cr a/lib/IR/Value.cpp b/lib/IR/Value.cpp
--- a/lib/IR/Value.cpp	2016-12-07 13:47:32.000000000 -0800
+++ b/lib/IR/Value.cpp	2018-03-12 11:25:42.752625100 -0700
@@ -466,6 +466,11 @@
       V = GEP->getPointerOperand();
     } else if (Operator::getOpcode(V) == Instruction::BitCast ||
                Operator::getOpcode(V) == Instruction::AddrSpaceCast) {
+      // Do not process across AddrSpaceCast, since its result could
+      // be a nullptr depending on the meaning of src and dest addrspace.
+      if (Operator::getOpcode(V) == Instruction::AddrSpaceCast) {
+         return V;
+      }
       V = cast<Operator>(V)->getOperand(0);
     } else if (GlobalAlias *GA = dyn_cast<GlobalAlias>(V)) {
       if (StripKind == PSK_ZeroIndices || GA->isInterposable())
