#    Description                : completely turn off code-sinking in InstructionCombining.


diff -Naur --strip-trailing-cr a/lib/Transforms/InstCombine/InstructionCombining.cpp b/lib/Transforms/InstCombine/InstructionCombining.cpp
--- a/lib/Transforms/InstCombine/InstructionCombining.cpp	2017-10-23 09:48:44.308224955 -0400
+++ b/lib/Transforms/InstCombine/InstructionCombining.cpp	2017-10-23 09:49:06.404224089 -0400
@@ -2871,7 +2871,9 @@
     }
 
     // See if we can trivially sink this instruction to a successor basic block.
-    if (I->hasOneUse()) {
+    /*** turn off code-sinking ***/
+    /***  if (I->hasOneUse()) { ***/
+    if (0) {
       BasicBlock *BB = I->getParent();
       Instruction *UserInst = cast<Instruction>(*I->user_begin());
       BasicBlock *UserParent;
