#    Description                  : Universal Driver Enabling


diff -Naur --strip-trailing-cr a/lib/Support/Windows/Process.inc b/lib/Support/Windows/Process.inc
--- a/lib/Support/Windows/Process.inc	2017-10-27 09:51:01.756680018 -0400
+++ b/lib/Support/Windows/Process.inc	2017-10-27 09:51:43.873678368 -0400
@@ -232,7 +232,10 @@
   LongPath.set_size(Length);
   return ConvertAndPushArg(LongPath.data(), Args, Allocator);
 }
-
+//INTEL_INTERNAL START
+//Disabling this function as it has a conflict with Windows Universal API
+//CommandLineToArgvW is the windows call causing the conflict.
+#if 0
 std::error_code
 Process::GetArgumentVector(SmallVectorImpl<const char *> &Args,
                            ArrayRef<const char *>,
@@ -267,6 +270,7 @@
   LocalFree(UnicodeCommandLine);
   return ec;
 }
+#endif
 
 std::error_code Process::FixupStandardFileDescriptors() {
   return std::error_code();
